### DISCUSSION TITLE
Have an option to disable the creation of a local variable upon language change

### DISCUSSION THREAD

**Discussion by brabalan** [minad/jinx#139] [Ideas]
When switching languages writing emails in notmuch, I’m always asked if I want to save a local variable for the language, which does not make much sense. Could it be possible to have a configuration variable that controls this?

**Comment by minad** [comment]
Good point. 

However it seems that Notmuch sets a `buffer-file-name`, which means that the file is supposed to be saved. The idea is that in such cases, storing file-local variables (including the language) makes sense. See:

https://github.com/minad/jinx/blob/cd63c27767d655319f89432ed1feae1cae12cdab/jinx.el#L919-L923

I wonder why Notmuch sets this variable. In contrast, when I am in `message-mode` from Gnus, `buffer-file-name` is nil.

**Reply by brabalan** [reply]
(Sorry for not replying earlier, I was not notified of your reply.)

It is not notmuch that seems to set a `buffer-file-name`, but `message-mode`:
```emacs-lisp
  (unless buffer-file-name
    (message-set-auto-save-file-name))
```
Maybe gnus disable it after…

I saw the code you added in the case of a missing file name, I’m wondering if it could be made more customizable.

**Reply by minad** [reply]
I see. It could be that I have this message auto save feature disabled in my config. I can check later how this is done. Would this be an option for you too?

**Reply by brabalan** [reply]
I disabled it doing
```emacs-lisp
(setq message-auto-save-directory nil)
```
and it works.

There is some special code in message-mode that does not set a buffer file name if gnus is alive:
```emacs-lisp
    (if (gnus-alive-p)
	(setq message-draft-article
	      (nndraft-request-associate-buffer "drafts"))
```

**Comment by rdiaz02** [comment]
I am running into a similar situation, using mu4e.  But this is also an issue in text files I send other people where adding local variables does not make sense.  For now, I have solved it by brute force by removing parts of `jinx-languages` and writing

```
(defun rdu-jinx-change-dict (langs)
  "So that it does not ask additional questions about saving a
file-local variable when using mu4e (or whatever)."
  (interactive)
  (setq-local jinx-languages langs)
  (jinx--load-dicts)
  (jinx--cleanup)
  )
```

But I am not sure if this is a good idea or if there is a better way. 

More generally, and since I think there are possibly many cases where adding local variables would not be a good idea, would it be possible to disable asking for saving, as another option of `jinx-languages`?

**Reply by minad** [reply]
> More generally, and since I think there are possibly many cases where adding local variables would not be a good idea, would it be possible to disable asking for saving, as another option of jinx-languages?

Could you please give me examples of such cases? If you sent text files to other people I don't see a reason why file local variables should not be present. They may be using Emacs as well. ;) The notmuch or mu4e cases should be solvable by setting buffer-file-name to nil.

**Reply by rdiaz02** [reply]
Any situation where the receiver is not (yet? ;-)) an Emacs user (and will, thus, not understand the meaning).  Some common cases for me:

- Sending files by email to anyone who is not using Emacs.
- Writing a text in Emacs that is then copied into a Whatsapp or Telegram or ... whatever.
- Editing text (in Emacs, of course) that ends up pasted in a libreoffice document or other similar things. 
- Editing text using emacs-anywhere or atomic-chrome (that ends up copied to github, reddit, any other web form, a libreoffice document, etc, etc), and making temporaty saves in a file while doing the editing.

Finally, I am not sure setting 'buffer-file-name'  to nil, at least with mu4e, is always what one wants. I might be wrong, but I think that might preclude creating temporary saves, with a unique name, of the draft.

**Reply by minad** [reply]
I am a bit bold here, but I don't see a strong argument against file-local variables. :)

> Sending files by email to anyone who is not using Emacs.

I often receive files in formats which I do not like, for example Word documents. So why should people be bothered by file local variables which are hidden behind a comment anyway (for example % in TeX)? This seems far less worse to me. Editors and their special mode often add such variables, for example AUCTeX.

> Writing a text in Emacs that is then copied into a Whatsapp or Telegram or ... whatever.

Why not do this in a temporary scratch-like buffer which is not backed by a file?

> Editing text (in Emacs, of course) that ends up pasted in a libreoffice document or other similar things.

Same.

> Editing text using emacs-anywhere or atomic-chrome (that ends up copied to github, reddit, any other web form, a libreoffice document, etc, etc), and making temporaty saves in a file while doing the editing.

It could happen that one writes a temporary note file (e.g. in Markdown format), which one then wants to open again. But at this point the file-local variable will already become beneficial.

**Reply by minad** [reply]
Anyway, in order to disable this feature, it should suffice to locally let-bind `buffer-file-name` to nil in an advice around the `jinx-languages` command.

**Reply by rdiaz02** [reply]
> I am a bit bold here

I'll try to match the boldness in my respone ;-)

> I often receive files in formats which I do not like, for example Word documents.

I fail to see how that is an argument. I, too, receive files in formats which I do not like. But precisely because that umistakeably makes me realize that the action is dislikeable, I try to avoid doing it to other people. 

More specifically for this case: I try to minimize surprise to the receiver because I think that communication gains when we eliminate (what the receiver probably perceives as) extraneous or unintelligible elements.  

> which are hidden behind a comment anyway (for example % in TeX)?
> their special mode often add such variables, for example AUCTeX

I can expect users of LaTeX (whether using AUCTeX or other things) to filter that; they filter lots of "%%".  But when emailing/whatsapping/etc  people who are not familiar with the notion of "comment" (I think they are the majority, almost surely among adolescentes and people over 80), I definitely think that  this

```
;; Local Variables:
;; jinx-languages: "es_ES"
;; End:
```
does not help communication.  They can easily wonder what that means and if they are misunderstanding something important: what is that local variable? Is that jinx something they should be concerned about? Etc. 

So, to summarize up to here: since I think there is potential for making communication harder and I only see negligible gains, why do it? 

> > Editing text (in Emacs, of course) that ends up pasted in a libreoffice document or other similar things.
> Same.

I don't think it is the same as the above examples. If it is part of, say, a paper one is writing with collaborators that are using libreoffice those three last lines must go. Regardless of the merits of my arguments for the examples above, here I do not see how one would argue that leaving those three lines here is inconsequential.

> It could happen that one writes a temporary note file (e.g. in Markdown format), which one then wants to open again. But at this point the file-local variable will already become beneficial.

Yes but: the saving in a file is, often,  "just in case of disaster". So most of the time one does not open the file again. However, all of the time one needs to, at least once, delete the last three lines.

 > Anyway, in order to disable this feature, it should suffice to locally let-bind buffer-file-name to nil in an advice around the jinx-languages command.

OK, thanks. Along those lines, I guess that my code above is nothing that will create unnecessary harm either, right? (Seemed simpler to me than adding advice).

**Reply by minad** [reply]
> I'll try to match the boldness in my respone ;-)

That's the spirit :)

> OK, thanks. Along those lines, I guess that my code above is nothing that will create unnecessary harm either, right? (Seemed simpler to me than adding advice).

It is not causing any harm. There are many ways to achieve your goals in Emacs, advices, custom commands, and so on.

**Reply by rdiaz02** [reply]
Thanks!

**Comment by minad** [comment]
After this feature has been requested once more, it got added in https://github.com/minad/jinx/commit/43b2e6413648ef437f297bc30f8830623cf4b266.

**Reply by brabalan** [reply]
Thank you!

**Reply by rdiaz02** [reply]
Thanks!
